packages/ai-constructs/src/conversation/conversation_handler_construct.ts

```diff
@@ -0,0 +1,104 @@
+import { FunctionResources, ResourceProvider } from '@aws-amplify/plugin-types';
+import { Duration } from 'aws-cdk-lib';
+import { Effect, PolicyStatement } from 'aws-cdk-lib/aws-iam';
+import { CfnFunction, Runtime as LambdaRuntime } from 'aws-cdk-lib/aws-lambda';
+import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
+import {
+  CustomDataIdentifier,
+  DataProtectionPolicy,
+  LogGroup,
+  RetentionDays,
+} from 'aws-cdk-lib/aws-logs';
+import { Construct } from 'constructs';
+import path from 'path';
+
+const resourcesRoot = path.normalize(path.join(__dirname, 'runtime'));
+const defaultHandlerFilePath = path.join(resourcesRoot, 'default_handler.js');
+
+export type ConversationHandlerFunctionProps = {
+  entry?: string;
+  models: Array<{
+    modelId: string;
+    region: string;
+  }>;
+};
+
+/**
+ * Conversation Handler Function CDK construct.
+ * This construct deploys resources that integrate conversation routes
+ * defined in data schema with AI models available in AWS Bedrock. I.e.
+ * 1. AWS Lambda function that handles conversation turn events.
+ *    With Amplify provided implementation by default and option to specify
+ *    custom handler.
+ * 2. AWS CloudWatch log group policy with appropriate data protection policies.
+ * 3. AWS IAM policy that grants access to selected AWS Bedrock models.
+ */
+export class ConversationHandlerFunction
+  extends Construct
+  implements ResourceProvider<FunctionResources>
+{
+  resources: FunctionResources;
+
+  /**
+   * Creates Conversation Handler Function CDK construct.
+   */
+  constructor(
+    scope: Construct,
+    id: string,
+    private readonly props: ConversationHandlerFunctionProps
+  ) {
+    super(scope, id);
+
+    if (this.props.entry && !path.isAbsolute(this.props.entry)) {
+      throw new Error('Entry must be absolute path');
+    }
+
+    const conversationHandler = new NodejsFunction(
+      this,
+      `conversationHandlerFunction`,
+      {
+        runtime: LambdaRuntime.NODEJS_18_X,
+        timeout: Duration.seconds(60),
+        entry: this.props.entry ?? defaultHandlerFilePath,
+        handler: 'handler',
+        bundling: {
+          bundleAwsSDK: true,
+        },
+        logGroup: new LogGroup(this, 'conversationHandlerFunctionLogGroup', {
+          retention: RetentionDays.INFINITE,
+          dataProtectionPolicy: new DataProtectionPolicy({
+            identifiers: [
+              new CustomDataIdentifier(
+                'JWTToken',
+                'ey[A-Za-z0-9-_=]+\\.[A-Za-z0-9-_=]+\\.?[A-Za-z0-9-_.+/=]*'
+              ),
+            ],
+          }),
+        }),
+      }
+    );
+
+    if (this.props.models && this.props.models.length > 0) {
+      const resources = this.props.models.map(
+        (model) =>
+          `arn:aws:bedrock:${model.region}::foundation-model/${model.modelId}`
```

Comment:
What happens if a customer's account is in one region and they use a model in a different region? Or is that not possible/not a concern?

Reply:
> if a customer's account

I guess you mean main stack resources. Account has access to many regions.

> they use a model in a different region?

That is perfectly legal. It just ends up being cross region call which comes with extra latency. The API is created in such a way that they have to opt in to do this and upstream layers (data construct?) will do some defaulting.
It is possible to allow usage of same model in multiple regions, one just adds another entry with same modelid and different region.
The reason to expose region at all as an option is that number of models available per region varies and is quite constrained comparing to other resources we provision (like cognito).

Reply:
Not in scope of this PR but maybe we can add a warning that there will be extra latency. Might be confusing if you switch models to one in another region and all of a sudden your conversation handler consistently times out.

Reply:
That latency penalty is not that high comparing to LLM execution time.

Take a look at this post https://repost.aws/questions/QUl7IRSjpMQVe2mMOFojP6qA/a-question-about-inter-region-latency to get a raw estimate of how impactful this can be. I.e. in case of S3 1MB transfers it seem to range between 81ms and 755ms depending on regions selections. Which is still order of magnitude less than LLM execution alone.

I'd say we should hold with such a warning as it's not something magical we'd be doing. Customer explicitly specifies that, so the fact that cross region communication may be costly should be rather obvious.

Reply:
Got it, wasn't sure if the extra latency would be substantial or not compared to the total execution time.